### PR TITLE
Allow invalid AMP creatives to get position info

### PR DESCRIPTION
**Pull request by zombifier**
Allows unregistered iframes (and their children) to receive position info from the inabox host script.

Some AMP creatives will fail validation, meaning it will not be registered with the host script and thus unable to determine its visibility or resize itself. The former functionality should be unrestricted to any frames that request it, valid or not.

### CODE REVIEW

**Comment by jeffkaufman** [comment]
@torch2424 Some context:

* When an AMP creative fails validation it won't be allowed to talk to the host script, which means if its in a cross-domain iframe on Safari (no IntersectionObserver) it can't determine its visibility.  This means it can't determine if it has met the IAB standard for viewability (50% on screen for a second).

* Creatives can fail validation for many reasons, including things like "the creative had 19k of CSS, and then the ad network added 5k of CSS, and so it's over the 20k limit".  Of course the creative could also have custom JS or just not be AMP at all.

* While some host script features do need to be restricted to validated AMP creatives (resizing) this isn't the case for position information, so it's ok to open the host script up to supply this information to any creative that requests it.

An alternative implementation of this would be for the ad tag to add all creatives, AMP or not, to the list of frames, and use the new `amp-allowed` whitelist (PR #18698) to restrict them to position information only.  I think the change here is simpler, but could be convinced otherwise.

@zombifier could you include some of this background in your PR description?

File: ads/inabox/inabox-messaging-host.js [hunk 1]
```diff
@@ -125,10 +125,19 @@ export class InaboxMessagingHost {
...
         this.getFrameElement_(message.source, request['sentinel']);
     if (!iframe) {
-      dev().info(TAG, 'Ignored message from untrusted iframe:', message);
+      dev().info(TAG, 'Ignored message from unmeasurable iframe:', message);
       return false;
     }
 
+    // if message comes from an unregistered frame, allow anyway if it's only
+    // getting viewability
+    if (!this.iframeMap_[request['sentinel']].iframe) {
```
Comment (jeffkaufman) [inline ads/inabox/inabox-messaging-host.js]: I'm not convinced this works: after running `getFrameElement_` above either:

* `getFrameElement_` returned `null`, `!iframe` was `true`, we returned `false` above
* `!this.iframeMap_[request['sentinel']].iframe` will be false and this code block will never run

Have you tested this locally?

Comment (zombifier) [reply ads/inabox/inabox-messaging-host.js]: getFrameElement_ does not return the iframe, it returns the top most measurable frame (measurableFrame), so the two conditions are not mutually exclusive. I had tested this locally and it works on my end.

Comment (jeffkaufman) [reply ads/inabox/inabox-messaging-host.js]: I'm still not seeing it, but maybe I'm misreading?

Case 1:

* `getFrameElement_` returns `null`
* `iframeMap_[sentinel]` is unset
* `processMessage` will log `Ignored message from unmeasurable iframe` and return false

Case 2:

* `getFrameElement_` returns non-`null`
* `iframeMap_[sentinel]` will be set, as will `iframeMap_[sentinel].iframe`
* `!this.iframeMap_[request['sentinel']].iframe` will be false
* the `if (request['type'] != MessageType.SEND_POSITIONS) {...}` block won't ever run

Are you getting `Ignored message from untrusted iframe` in local testing?

**Comment by lannka** [comment]
Not sure if it's the same thing but I remember we were discussing adding `amp-allow` attribute to inform host.js what APIs are granted to the ad.

Is that the same thing?

@zombifier @jeffkaufman since you've already written up a description, do you mind moving it to an issue I2I, and have this PR link to it. Just for streamlined tracking and providing more background to the community

**Comment by zombifier** [comment]
So concerning @jeffkaufman's comment, after a bit of thinking and discussions I believe that the stated alternative of loosening the ad tag's restrictions so all frames are registered with the host script and instead rely on the whitelist is the better way to do this. As such, I'm closing this PR for now.

**Pull request closed without merging**